lib/ustr2stp.h

```diff
@@ -0,0 +1,65 @@
+/*
+ * SPDX-FileCopyrightText: 2022-2023, Alejandro Colomar <alx@kernel.org>
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+
+#ifndef SHADOW_INCLUDE_LIBMISC_USTR2STP_H_
+#define SHADOW_INCLUDE_LIBMISC_USTR2STP_H_
+
+
+#include <config.h>
+
+#include <assert.h>
+#include <stddef.h>
+#include <string.h>
+
+#include "must_be.h"
+#include "sizeof.h"
+
+
+inline char *ustr2stp(char *restrict dst, const char *restrict src,
+    size_t ssize);
+
+
+/*
+ * SYNOPSIS
+ *	char *ustr2stp(char *restrict dst,
+ *	               const char src[restrict .ssize], size_t ssize);
+ *
+ * ARGUMENTS
+ *	dst	Destination buffer.
+ *	src	Source character sequence.
+ *	ssize	Size of the source character sequence.
+ *
+ * DESCRIPTION
+ *	This function copies 'ssize' bytes from src to dst, and places a
+ *	terminating null byte after that.  It reads a character
+ *	sequence, and produces a string.
+ *
+ * RETURN VALUE
+ *	dst + ssize
+ *		This function returns a pointer to the terminating null
+ *		byte.
+ *
+ * CAVEATS
+ *	This function doesn't know the size of the destination buffer.
+ *	It assumes it will always be large enough.  Since the size of
+ *	the source buffer is known to the caller, it should make sure to
+ *	allocate a destination buffer of at least `ssize + 1`.
+ */
+
+
+inline char *
+ustr2stp(char *restrict dst, const char *restrict src, size_t ssize)
+{
+	char  *p;
+
+	p = mempcpy(dst, src, ssize);
+	*p = '\0';
```

Comment:
Do you think it's worth saying "if (p != dst) *p = '\0';" ?

Reply:
That would be wrong.  `ssize` is the source size.  The size in `dst` should always be `ssize + 1`, which guarantees that there's always room for the `'\0'`.

Imagine I want to write an empty string with this function:

```c
ustr2stp(dst, "", 0);
```

That would write a `'\0'` at `dst`.

Reply:
> Or maybe that should just be a BUG_ON(size == 0)

It should rather be a `BUG_ON(sizeof(dst) <= ssize)`

Reply:
Can we make ustr2stp() be a #define that first checks the dst != NULL and NITEMS(dst) > 0?  I don't really care about NITEMS(dst) <= ssize if ssize > 0, because it won't break anything.  I only care about us avoiding trying to set an invalid address to '\0'.

Reply:
v2:

```c
$ grepc -tfd ustr2stp .
./lib/ustr2stp.h:inline char *
ustr2stp(char *restrict dst, const char *restrict src, size_t ssize)
{
	return strcpy(mempcpy(dst, src, ssize), "");
}
```

Reply:
>  In shadow, there are no zero-sized arrays at all

I worry about future changes.

And this is where introducing all of the new functions can be much worse than consolidating on using one less-than-ideal function.  If we were to use snprintf everywhere and always check for rv, then everyone would know what to expect.  Few contributors and reviewers will remember all of the edge cases with all the new functions.

My hope is that your explanations about zero padded vs null terminated strings will help with this.  But the function names are, frankly attrocious.  Probably every function description should remind the reader 'ustr means x, stp means y'.

But this patch looks good, so I'll merge it now, thanks.